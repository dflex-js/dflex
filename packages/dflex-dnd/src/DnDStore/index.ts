export { default as store } from "./DFlexDnDStore";
export { default as initDFlexEvent } from "./DFlexEvents";
export { default as initDFlexListeners } from "./DFlexListeners";

export type { ElmTree } from "./types";

export type {
  DFlexDraggedEvent,
  DFlexInteractivityEvent,
  DFlexSiblingsEvent,
  DFlexEvents,
  DFlexEventsTypes,
  DFlexEventsMap,
  DFlexEventPlugin as DFlexEventInitializer,
} from "./DFlexEvents";

export type {
  LayoutState as DFlexLayoutState,
  DFlexLayoutStateEvent,
} from "./DFlexListeners";
