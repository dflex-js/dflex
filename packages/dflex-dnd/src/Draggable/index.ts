// eslint-disable-next-line import/prefer-default-export
export { default as DraggableInteractive } from "./DraggableInteractive";
