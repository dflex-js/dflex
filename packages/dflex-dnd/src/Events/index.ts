export { default as DFlexEvent } from "./DFlexEvents";

export {
  createInteractivityPayload,
  createSiblingsPayload,
  createDragPayload,
} from "./payloads";

export { DFLEX_EVENTS, DFLEX_EVENTS_CAT } from "./constants";

export type {
  DFlexDraggedEvent,
  DFlexInteractivityEvent,
  DFlexSiblingsEvent,
  DFlexEvents,
  DFlexEventNames as DFlexEventsTypes,
  DFlexEventsMap,
} from "./types";
