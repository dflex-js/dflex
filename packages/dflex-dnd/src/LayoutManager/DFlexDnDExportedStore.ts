import type { DFlexGlobalConfig, RegisterInputOpts } from "@dflex/store";
import type { DFlexSerializedElement } from "@dflex/core-instance";
import type { Siblings } from "@dflex/dom-gen";

import DFlexDnDStore from "./DFlexDnDStore";

import { store } from ".";

// TODO: I am not sure if this is the best approach or not.

class DFlexDnDExportedStore {
  private _base: DFlexDnDStore;

  listeners: DFlexDnDStore["listeners"];

  constructor() {
    this._base = store;

    this.listeners = this._base.listeners;
  }

  /**
   * Sets DFlex global configurations.
   *
   * @param globals - Global configurations for DFlex.
   */
  config(globals: DFlexGlobalConfig): void {
    this._base.config(globals);
  }

  /**
   * Checks if an element with the specified ID is registered.
   *
   * @param id - ID of the element.
   * @returns `true` if the element is registered, `false` otherwise.
   */
  has(id: string): boolean {
    return this._base.has(id);
  }

  /**
   * Retrieves all element IDs of siblings in the given node represented by the sibling key.
   *
   * @param SK - Sibling Key.
   * @returns An object containing the sibling element IDs.
   */
  getElmSiblingsByKey(SK: string): Siblings {
    return this._base.getElmSiblingsByKey(SK);
  }

  /**
   * Checks if the DFlex instance is idle.
   *
   * @returns `true` if DFlex is idle, `false` otherwise.
   */
  isIDle(): boolean {
    return this._base.isIDle();
  }

  /**
   * Checks if the layout is available.
   *
   * @returns `true` if the layout is available, `false` otherwise.
   */
  isLayoutAvailable(): boolean {
    return this._base.isLayoutAvailable();
  }

  /**
   * Registers an element with DFlex.
   *
   * @param elm - Element to register.
   * @returns The registered element.
   */
  register(elm: RegisterInputOpts) {
    return this._base.register(elm);
  }

  commit(): void {
    this._base.commit(null);
  }

  /**
   * Retrieves the serialized representation of the element with the specified ID.
   *
   * @param id - ID of the element.
   * @returns The serialized element object or `null` if the element is not found.
   */
  getSerializedElm(id: string): DFlexSerializedElement | null {
    return this.getSerializedElm(id);
  }

  /**
   * Destroys the DFlex instance.
   */
  destroy() {
    return this._base.destroy();
  }
}

export default DFlexDnDExportedStore;
