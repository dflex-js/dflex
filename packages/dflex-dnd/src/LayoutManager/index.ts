export { default as store } from "./DFlexDnDStoreSingleton";
export { default as initDFlexEvent } from "./DFlexEvents";
export { default as initDFlexListeners } from "./DFlexListeners";
export { default as scheduler } from "./DFlexScheduler";
export { addMutationObserver as initMutationObserver } from "./DFlexMutations";

export type {
  DFlexDraggedEvent,
  DFlexInteractivityEvent,
  DFlexSiblingsEvent,
  DFlexEvents,
  DFlexEventsTypes,
  DFlexEventPlugin,
} from "./DFlexEvents";

export type {
  DFlexListenerEvents,
  DFlexLayoutStateEvent,
  DFlexElmMutationEvent,
} from "./DFlexListeners";

export { DFLEX_EVENTS } from "./constants";
