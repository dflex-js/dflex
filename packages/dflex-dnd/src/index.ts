export { exportedStoreSingleton as store } from "./store";

export { default as DnD } from "./DnD";

export type { DFlexDnDOpts } from "./types";

export type {
  DFlexDraggedEvent,
  DFlexInteractivityEvent,
  DFlexSiblingsEvent,
  DFlexEvents,
  DFlexEventsTypes,
  DFlexLayoutStateEvent,
  DFlexElmMutationEvent,
  DFlexListenerEvents,
} from "./LayoutManager";

export type {
  DFlexScrollContainer,
  DFlexSerializedElement,
  DFlexDOMGenOrder,
} from "@dflex/core-instance";

export type { RegisterInputOpts, DFlexGlobalConfig } from "@dflex/store";

export type {
  AxesPoint,
  AbstractBoxRect as BoxRectAbstract,
  PointNum,
  AbstractBox,
  Dimensions,
} from "@dflex/utils";
