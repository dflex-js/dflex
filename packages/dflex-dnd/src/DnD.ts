import type { AxesPoint } from "@dflex/utils";

import DraggableInteractive from "./Draggable";
import Droppable from "./Droppable";
import { scheduler, store } from "./LayoutManager";

import type { DFlexDnDOpts, FinalDndOpts } from "./types";

import { extractOpts } from "./utils/extractOpts";
import { defaultOpts } from "./utils/constants";

class DnD extends Droppable {
  /**
   *
   * @param id -
   * @param initCoordinates -
   * @param opts -
   */
  constructor(
    id: string,
    initCoordinates: AxesPoint,
    opts: DFlexDnDOpts = defaultOpts
  ) {
    if (__DEV__) {
      if (!store.registry.has(id)) {
        throw new Error(`DFlex: ${id} is not registered in the Store.`);
      }
    }

    const options = extractOpts(opts);

    const draggable = new DraggableInteractive(
      id,
      initCoordinates,
      options as FinalDndOpts
    );

    super(draggable);

    scheduler(store, null, null, { layoutState: "ready", type: "layoutState" });
  }
}

export default DnD;
