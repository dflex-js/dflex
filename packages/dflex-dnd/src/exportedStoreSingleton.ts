import { DFlexDnDExportedStore } from "./LayoutManager";

function createStoreInstance() {
  const store = DFlexDnDExportedStore.getInstance();

  return store;
}

export default createStoreInstance();
