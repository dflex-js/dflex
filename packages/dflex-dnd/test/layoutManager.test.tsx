/* eslint-disable import/no-extraneous-dependencies */
import * as React from "react";
import * as ReactTestUtils from "react-dom/test-utils";
import { Root, createRoot } from "react-dom/client";

import { PointNum } from "@dflex/utils";
import { store } from "../src/LayoutManager";

describe("DFlex DnD", () => {
  const elm1 = {
    id: "id-1",
    depth: 0,
    readonly: false,
  };

  const elm2 = {
    id: "id-2",
    depth: 0,
    readonly: false,
  };

  const elm3 = {
    id: "id-3",
    depth: 0,
    readonly: false,
  };

  const elm4 = {
    id: "id-4",
    depth: 0,
    readonly: false,
  };

  let container: HTMLDivElement | null;
  let reactRoot: Root;

  const positions: Record<string, DOMRect> = {};

  beforeAll(() => {
    jest.useFakeTimers();
    window.HTMLElement.prototype.scrollIntoView = jest.fn();

    container = document.createElement("div");
    reactRoot = createRoot(container);
    document.body.appendChild(container);

    function DnDComponent(elm: typeof elm3) {
      const ref = React.createRef<HTMLDivElement>();

      React.useEffect(() => {
        if (ref.current) {
          store.register(elm);

          positions[elm.id] = ref.current.getBoundingClientRect();
        }

        return () => {
          store.unregister(elm.id);
        };
      }, [ref]);

      return <div id={elm.id} key={elm.id} ref={ref} />;
    }

    function init() {
      function TestBase() {
        const ref = React.createRef<HTMLDivElement>();

        return (
          <div ref={ref}>
            <div>
              <DnDComponent {...elm1} />
            </div>
            <div>{[elm2, elm3, elm4].map((elm) => DnDComponent(elm))}</div>
          </div>
        );
      }

      ReactTestUtils.act(() => {
        reactRoot.render(<TestBase />);
      });
    }

    init();
    jest.runAllTimers();
  });

  afterAll(() => {
    document.body.removeChild(container!);
    container = null;
  });

  describe("DFlex DnD Store registration", () => {
    it("Element is initiated", () => {
      store.register(elm1);
      expect(store).toMatchSnapshot();
    });

    describe("Checking the two branches, depth and keys", () => {
      it("Store has two branches", () => {
        expect(store.getBranchesByDepth(0).length).toBe(2);
      });

      it("Store depth zero matches branches Siblings key(SK)", () => {
        const branches = store.getBranchesByDepth(0);

        expect(branches[0]).toBe(store.registry.get(elm1.id)!.keys.SK);
        expect(branches[1]).toBe(store.registry.get(elm2.id)!.keys.SK);
      });
    });

    describe("Checking containers", () => {
      it("Store has two containers matching the two branches", () => {
        expect(store.containers.size).toBe(2);
      });

      it("Containers have the correct origin length", () => {
        const container1 = store.containers.get(
          store.registry.get(elm1.id)!.keys.SK
        )!;

        expect(container1.originLength).toBe(1);

        const container2 = store.containers.get(
          store.registry.get(elm2.id)!.keys.SK
        )!;

        expect(container2.originLength).toBe(3);
      });

      it.skip("Containers initialized `lastElmPosition`", () => {
        const container1 = store.containers.get(
          store.registry.get(elm1.id)!.keys.SK
        )!;

        expect(container1.lastElmPosition).toBeInstanceOf(PointNum);

        const container2 = store.containers.get(
          store.registry.get(elm2.id)!.keys.SK
        )!;

        expect(container2.lastElmPosition).toBeInstanceOf(PointNum);
      });
    });

    describe("Checking interactiveDOM", () => {
      it("interactiveDOM has all elements DOM", () => {
        [elm1, elm2, elm3, elm4].forEach((elm) => {
          expect(store.interactiveDOM.has(elm.id)).toBe(true);
        });
      });

      it("interactiveDOM has also the parent containers", () => {
        expect(store.interactiveDOM.size).toBe(6);
      });

      it("Elements have the right dataset", () => {
        [elm1].forEach((elm, i) => {
          expect(store.interactiveDOM.get(elm.id)!.dataset.index).toBe(`${i}`);
        });

        [elm2, elm3, elm4].forEach((elm, i) => {
          expect(store.interactiveDOM.get(elm.id)!.dataset.index).toBe(`${i}`);
        });
      });
    });
  });
});
