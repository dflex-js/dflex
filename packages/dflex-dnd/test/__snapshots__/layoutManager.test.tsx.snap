// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DFlex DnD DFlex DnD Store registration Element is initiated 1`] = `
DFlexDnDStore {
  "DOMGen": DOMKeysGenerator {
    "_PKByDepth": {
      "0": "dflex_ky_1_1",
      "1": "dflex_ky_2_0",
    },
    "_SKByDepth": Map {
      0 => [
        "dflex_sk_0_0",
        "dflex_sk_0_1",
      ],
      1 => [
        "dflex_sk_1_0",
      ],
    },
    "_branchIndex": 1,
    "_branchesRegistry": Map {
      "dflex_bk_0" => Map {
        0 => {
          "SK": "dflex_sk_0_0",
          "ids": [
            "id-1",
          ],
        },
        1 => {
          "SK": "dflex_sk_1_0",
          "ids": [
            "dflex_id_0",
          ],
        },
      },
      "dflex_bk_1" => Map {
        0 => {
          "SK": "dflex_sk_0_1",
          "ids": [
            "id-2",
            "id-3",
            "id-4",
          ],
        },
        1 => {
          "SK": "dflex_sk_1_0",
          "ids": [
            "dflex_id_1",
          ],
        },
      },
    },
    "_idsBySk": Map {
      "dflex_sk_0_0" => [
        "id-1",
      ],
      "dflex_sk_1_0" => [
        "dflex_id_0",
        "dflex_id_1",
      ],
      "dflex_sk_0_1" => [
        "id-2",
        "id-3",
        "id-4",
      ],
    },
    "_preBK": null,
    "_prevBKs": [
      "dflex_bk_0",
      "dflex_bk_1",
    ],
    "_prevDepth": 1,
    "_prevPK": "dflex_ky_2_0",
    "_siblingsCount": {
      "0": 4,
      "1": 2,
      "2": 0,
    },
  },
  "_initObservers": [Function],
  "_initSiblings": [Function],
  "_isDOM": true,
  "_isInitialized": true,
  "_lastDOMParent": null,
  "_resizeThrottle": [Function],
  "_taskQ": TaskQueue {
    "_elmInQueue": Set {},
    "_queue": {},
  },
  "_terminatedDOMiDs": Set {},
  "_unregisterSchedule": [Function],
  "_windowResizeHandler": [Function],
  "containers": Map {
    "dflex_sk_0_0" => DFlexParentContainer {
      "_boundariesByRow": BoxNum {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "_gridIndex": PointNum {
        "x": 0,
        "y": 0,
      },
      "_rect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "_siblingBoundaries": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "grid": PointNum {
        "x": 0,
        "y": 0,
      },
      "id": "dflex_id_0",
      "lastElmPosition": null,
      "originLength": 1,
    },
    "dflex_sk_0_1" => DFlexParentContainer {
      "_boundariesByRow": BoxNum {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "_gridIndex": PointNum {
        "x": 0,
        "y": 2,
      },
      "_rect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "_siblingBoundaries": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "grid": PointNum {
        "x": 0,
        "y": 2,
      },
      "id": "dflex_id_1",
      "lastElmPosition": null,
      "originLength": 3,
    },
  },
  "deferred": [],
  "deletedDOM": WeakSet {},
  "globals": {
    "enableDragSettleOnSwitch": true,
    "enableEvents": true,
    "enableListeners": true,
    "removeEmptyContainer": false,
  },
  "interactiveDOM": Map {
    "id-1" => <div
      data-dev-x="0"
      data-dev-y="0"
      data-index="0"
      id="id-1"
    />,
    "dflex_id_0" => <div
      data-dflex-key="dflex_sk_1_0"
      data-index="0"
      id="dflex_id_0"
      style="opacity: 1;"
    >
      <div
        data-dev-x="0"
        data-dev-y="0"
        data-index="0"
        id="id-1"
      />
    </div>,
    "id-2" => <div
      data-dev-x="0"
      data-dev-y="0"
      data-index="0"
      id="id-2"
    />,
    "id-3" => <div
      data-dev-x="0"
      data-dev-y="1"
      data-index="1"
      id="id-3"
    />,
    "id-4" => <div
      data-dev-x="0"
      data-dev-y="2"
      data-index="2"
      id="id-4"
    />,
    "dflex_id_1" => <div
      data-dflex-key="dflex_sk_1_0"
      data-index="1"
      id="dflex_id_1"
      style="opacity: 1;"
    >
      <div
        data-dev-x="0"
        data-dev-y="0"
        data-index="0"
        id="id-2"
      />
      <div
        data-dev-x="0"
        data-dev-y="1"
        data-index="1"
        id="id-3"
      />
      <div
        data-dev-x="0"
        data-dev-y="2"
        data-index="2"
        id="id-4"
      />
    </div>,
  },
  "isComposing": true,
  "isUpdating": false,
  "listeners": {
    "clear": [Function],
    "notify": [Function],
    "subscribe": [Function],
  },
  "migration": null,
  "mutationObserverMap": Map {
    "dflex_id_0" => MutationObserver {},
    "dflex_id_1" => MutationObserver {},
  },
  "pending": [],
  "registry": Map {
    "id-1" => DFlexElement {
      "DOMGrid": PointNum {
        "x": 0,
        "y": 0,
      },
      "DOMOrder": {
        "parent": 0,
        "self": 0,
      },
      "VDOMOrder": {
        "parent": 0,
        "self": 0,
      },
      "_CSSTransform": null,
      "_RAF": [Function],
      "_animation": {
        "duration": "dynamic",
        "easing": "ease-in",
      },
      "_computedDimensions": null,
      "_hasPendingTransform": false,
      "_initialPosition": PointNum {
        "x": 0,
        "y": 0,
      },
      "_isVisible": true,
      "_translateHistory": undefined,
      "depth": 0,
      "id": "id-1",
      "keys": {
        "BK": "dflex_bk_0",
        "CHK": null,
        "PK": "dflex_ky_1_0",
        "SK": "dflex_sk_0_0",
      },
      "readonly": false,
      "rect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "translate": PointNum {
        "x": 0,
        "y": 0,
      },
    },
    "dflex_id_0" => DFlexElement {
      "DOMGrid": PointNum {
        "x": 0,
        "y": 0,
      },
      "DOMOrder": {
        "parent": 0,
        "self": 0,
      },
      "VDOMOrder": {
        "parent": 0,
        "self": 0,
      },
      "_CSSTransform": null,
      "_RAF": [Function],
      "_animation": {
        "duration": "dynamic",
        "easing": "ease-in",
      },
      "_computedDimensions": null,
      "_hasPendingTransform": false,
      "_initialPosition": PointNum {
        "x": 0,
        "y": 0,
      },
      "_isVisible": true,
      "_translateHistory": undefined,
      "depth": 1,
      "id": "dflex_id_0",
      "keys": {
        "BK": "dflex_bk_0",
        "CHK": "dflex_ky_1_0",
        "PK": "dflex_ky_2_0",
        "SK": "dflex_sk_1_0",
      },
      "readonly": true,
      "rect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "translate": PointNum {
        "x": 0,
        "y": 0,
      },
    },
    "id-2" => DFlexElement {
      "DOMGrid": PointNum {
        "x": 0,
        "y": 0,
      },
      "DOMOrder": {
        "parent": 1,
        "self": 0,
      },
      "VDOMOrder": {
        "parent": 1,
        "self": 0,
      },
      "_CSSTransform": null,
      "_RAF": [Function],
      "_animation": {
        "duration": "dynamic",
        "easing": "ease-in",
      },
      "_computedDimensions": null,
      "_hasPendingTransform": false,
      "_initialPosition": PointNum {
        "x": 0,
        "y": 0,
      },
      "_isVisible": true,
      "_translateHistory": undefined,
      "depth": 0,
      "id": "id-2",
      "keys": {
        "BK": "dflex_bk_1",
        "CHK": null,
        "PK": "dflex_ky_1_1",
        "SK": "dflex_sk_0_1",
      },
      "readonly": false,
      "rect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "translate": PointNum {
        "x": 0,
        "y": 0,
      },
    },
    "id-3" => DFlexElement {
      "DOMGrid": PointNum {
        "x": 0,
        "y": 1,
      },
      "DOMOrder": {
        "parent": 1,
        "self": 1,
      },
      "VDOMOrder": {
        "parent": 1,
        "self": 1,
      },
      "_CSSTransform": null,
      "_RAF": [Function],
      "_animation": {
        "duration": "dynamic",
        "easing": "ease-in",
      },
      "_computedDimensions": null,
      "_hasPendingTransform": false,
      "_initialPosition": PointNum {
        "x": 0,
        "y": 0,
      },
      "_isVisible": true,
      "_translateHistory": undefined,
      "depth": 0,
      "id": "id-3",
      "keys": {
        "BK": "dflex_bk_1",
        "CHK": null,
        "PK": "dflex_ky_1_1",
        "SK": "dflex_sk_0_1",
      },
      "readonly": false,
      "rect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "translate": PointNum {
        "x": 0,
        "y": 0,
      },
    },
    "id-4" => DFlexElement {
      "DOMGrid": PointNum {
        "x": 0,
        "y": 2,
      },
      "DOMOrder": {
        "parent": 1,
        "self": 2,
      },
      "VDOMOrder": {
        "parent": 1,
        "self": 2,
      },
      "_CSSTransform": null,
      "_RAF": [Function],
      "_animation": {
        "duration": "dynamic",
        "easing": "ease-in",
      },
      "_computedDimensions": null,
      "_hasPendingTransform": false,
      "_initialPosition": PointNum {
        "x": 0,
        "y": 0,
      },
      "_isVisible": true,
      "_translateHistory": undefined,
      "depth": 0,
      "id": "id-4",
      "keys": {
        "BK": "dflex_bk_1",
        "CHK": null,
        "PK": "dflex_ky_1_1",
        "SK": "dflex_sk_0_1",
      },
      "readonly": false,
      "rect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "translate": PointNum {
        "x": 0,
        "y": 0,
      },
    },
    "dflex_id_1" => DFlexElement {
      "DOMGrid": PointNum {
        "x": 0,
        "y": 0,
      },
      "DOMOrder": {
        "parent": 0,
        "self": 1,
      },
      "VDOMOrder": {
        "parent": 0,
        "self": 1,
      },
      "_CSSTransform": null,
      "_RAF": [Function],
      "_animation": {
        "duration": "dynamic",
        "easing": "ease-in",
      },
      "_computedDimensions": null,
      "_hasPendingTransform": false,
      "_initialPosition": PointNum {
        "x": 0,
        "y": 0,
      },
      "_isVisible": true,
      "_translateHistory": undefined,
      "depth": 1,
      "id": "dflex_id_1",
      "keys": {
        "BK": "dflex_bk_1",
        "CHK": "dflex_ky_1_1",
        "PK": "dflex_ky_2_0",
        "SK": "dflex_sk_1_0",
      },
      "readonly": true,
      "rect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "translate": PointNum {
        "x": 0,
        "y": 0,
      },
    },
  },
  "scrolls": Map {
    "dflex_sk_0_0" => DFlexScrollContainer {
      "_SK": "dflex_sk_0_0",
      "_containerDOM": <html>
        <head />
        <body>
          <div>
            <div>
              <div
                data-dflex-key="dflex_sk_1_0"
                data-index="0"
                id="dflex_id_0"
                style="opacity: 1;"
              >
                <div
                  data-dev-x="0"
                  data-dev-y="0"
                  data-index="0"
                  id="id-1"
                />
              </div>
              <div
                data-dflex-key="dflex_sk_1_0"
                data-index="1"
                id="dflex_id_1"
                style="opacity: 1;"
              >
                <div
                  data-dev-x="0"
                  data-dev-y="0"
                  data-index="0"
                  id="id-2"
                />
                <div
                  data-dev-x="0"
                  data-dev-y="1"
                  data-index="1"
                  id="id-3"
                />
                <div
                  data-dev-x="0"
                  data-dev-y="2"
                  data-index="2"
                  id="id-4"
                />
              </div>
            </div>
          </div>
        </body>
      </html>,
      "_isCandidateForDynamicVisibility": false,
      "_isDocumentContainer": true,
      "_listenerDatasetKey": "scroll__dflex_sk_0_0",
      "_scrollEventCallback": [Function],
      "_thresholdInViewport": {
        "inner": {
          "key": "scroll_inner_dflex_sk_0_0",
          "threshold": null,
        },
        "outer": {
          "key": "scroll_outer_dflex_sk_0_0",
          "threshold": null,
        },
      },
      "_throttledResizeHandler": [Function],
      "_throttledScrollHandler": [Function],
      "hasOverflow": PointBool {
        "x": false,
        "y": false,
      },
      "totalScrollRect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "visibleScrollRect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
    },
    "dflex_sk_0_1" => DFlexScrollContainer {
      "_SK": "dflex_sk_0_1",
      "_containerDOM": <html>
        <head />
        <body>
          <div>
            <div>
              <div
                data-dflex-key="dflex_sk_1_0"
                data-index="0"
                id="dflex_id_0"
                style="opacity: 1;"
              >
                <div
                  data-dev-x="0"
                  data-dev-y="0"
                  data-index="0"
                  id="id-1"
                />
              </div>
              <div
                data-dflex-key="dflex_sk_1_0"
                data-index="1"
                id="dflex_id_1"
                style="opacity: 1;"
              >
                <div
                  data-dev-x="0"
                  data-dev-y="0"
                  data-index="0"
                  id="id-2"
                />
                <div
                  data-dev-x="0"
                  data-dev-y="1"
                  data-index="1"
                  id="id-3"
                />
                <div
                  data-dev-x="0"
                  data-dev-y="2"
                  data-index="2"
                  id="id-4"
                />
              </div>
            </div>
          </div>
        </body>
      </html>,
      "_isCandidateForDynamicVisibility": false,
      "_isDocumentContainer": true,
      "_listenerDatasetKey": "scroll__dflex_sk_0_1",
      "_scrollEventCallback": [Function],
      "_thresholdInViewport": {
        "inner": {
          "key": "scroll_inner_dflex_sk_0_1",
          "threshold": null,
        },
        "outer": {
          "key": "scroll_outer_dflex_sk_0_1",
          "threshold": null,
        },
      },
      "_throttledResizeHandler": [Function],
      "_throttledScrollHandler": [Function],
      "hasOverflow": PointBool {
        "x": false,
        "y": false,
      },
      "totalScrollRect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
      "visibleScrollRect": BoxRect {
        "bottom": 0,
        "height": 0,
        "left": 0,
        "right": 0,
        "top": 0,
        "width": 0,
      },
    },
  },
  "unifiedContainerDimensions": {
    "0": {
      "height": 0,
      "width": 0,
    },
  },
  "updatesQueue": [],
}
`;
